src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs

```diff
@@ -159,7 +163,7 @@ public ImmutableArray<DiagnosticAsyncToken> ActiveDiagnosticTokens
         {
```

Comment:
HasPendingWork doesn't seem safe.  It reads shared mutable state without also ensuring an appropriate lock/gate to make sure reads/writes are appropriately fenced.